include/ck/utility/data_type.hpp

```diff
@@ -886,6 +922,385 @@ struct vector_type<T, 256>
     }
 };
 
+// Following vector_type specializations for int4_t is a hack to workaround
+// clang's lack of support for vector types of _BitInt(4).
+#ifdef CK_EXPERIMENTAL_BIT_INT_EXTENSION_INT4
+template <>
+struct vector_type<int4_t, 2>
+{
+    using d1_t = int4_t;
+    using d2_t = typename vector_type<int8_t, 1>::type;
+
+    using type = d2_t;
+
+    union
+    {
+        d2_t d2_;
+        StaticallyIndexedArray<d1_t, 2> d1x2_;
+        StaticallyIndexedArray<d2_t, 1> d2x1_;
+    } data_;
+
+    __host__ __device__ constexpr vector_type() : data_{type{0}} {}
+
+    __host__ __device__ constexpr vector_type(type v) : data_{v} {}
+
+    template <typename X>
+    __host__ __device__ constexpr const auto& AsType() const
+    {
+        static_assert(std::is_same_v<X, d1_t> || std::is_same_v<X, d2_t>, "wrong!");
+
+        if constexpr(std::is_same_v<X, d1_t>)
+        {
+            return data_.d1x2_;
+        }
+        else if constexpr(std::is_same_v<X, d2_t>)
+        {
+            return data_.d2x1_;
+        }
+    }
+
+    template <typename X>
+    __host__ __device__ constexpr auto& AsType()
+    {
+        static_assert(std::is_same_v<X, d1_t> || std::is_same_v<X, d2_t>, "wrong!");
+
+        if constexpr(std::is_same_v<X, d1_t>)
+        {
+            return data_.d1x2_;
+        }
+        else if constexpr(std::is_same_v<X, d2_t>)
+        {
+            return data_.d2x1_;
+        }
+    }
+};
+
+template <>
+struct vector_type<int4_t, 4>
+{
+    using d1_t = int4_t;
+    using d2_t = typename vector_type<int8_t, 1>::type;
+    using d4_t = typename vector_type<int8_t, 2>::type;
+
+    using type = d4_t;
+
+    union
+    {
+        d4_t d4_;
+        StaticallyIndexedArray<d1_t, 4> d1x4_;
```

Comment:
Should lines like `StaticallyIndexedArray<d1_t, *> d1x*`  be removed,  which causes the union size be doubled?

Reply:
Eagle eye! That's a good question. I need to check whether they're necessary.

Reply:
I guess they're needed. Take a look at those lines: 
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/718efa56537db121d1b19f944e6ef0451c30c277/include/ck/tensor_operation/gpu/thread/threadwise_tensor_slice_transfer_v3r1.hpp#L202-L216
`src_vector_type` is ie. `vector<int4_t, 16>` then in line 213 each vector's int4 element is accessed individually.

Reply:
I think what we need is to store  `16 int4_t` values using `8 int8_t`, but provide the capability to get each single int4_t value from the `StaticallyIndexedArray<int8_t, 8>`  and the specified index.

Reply:
Right now we do not need packing int4 data in int8. That's my mistake to implement it like this. Either way scalar int4 value occupies 1 byte, so we could express all mfma and memory operations using int8 instructions.